### ISSUE TITLE
Allow implementations without explicit interfaces

### ISSUE THREAD

**Issue by smithdtyler** [saeaadl/aadlv3#12]
Following the YAGNI principle, AADL should allow modelers to jump straight to writing an implementation of a component without declaring an interface (just as Java allows programmers to write a class without first writing an interface). 

Relaxing the requirement for a declared interface would make AADL less verbose and make it easier to quickly prototype models.

**Comment by joeseibel** [comment]
Here is an example of a way of declaring inline components:
```
package Nested_Subcomponents
public
	system s
	end s;

	system implementation s.i
		subcomponents
			ps1: process [
				features
					ps1_out: out data port;
				subcomponents
					t1: thread [
						features
							t1_out: out data port;
					];
				connections
					ps1_conn: port t1.t1_out -> ps1_out;
			];
			ps2: process [
				features
					ps2_in: in data port;
				subcomponents
					t2: thread [
						features
							t2_in: in data port;
					];
				connections
					ps2_conn: port ps2_in -> t2.t2_in;
			]
		connections
			across_conn: port ps1.ps1_out -> ps2.ps2_in;
	end s.i;
end Nested_Subcomponents;
```

**Comment by yoogx** [comment]
Quickly prototyping of models is an IDE issue, not a language one.

Here the concern is that you're building implicitly an instance model. I fail to see any improvements in readability when you stack also annexes in the definition.

Also, how would you name elements in the hierarchy, e.g. when you want to add flow ? are you applying is to the component instance or anonymous classifier? this proposal would impact name resolution. 

Note that some modelling notations also blur the distinction between component type and implementations, if you want to enter into this business of syntax reduction, we can even go further. 

But again, we have to be careful about simplification vs. strong typing.

**Comment by smithdtyler** [comment]
> Quickly prototyping of models is an IDE issue, not a language one.

Perhaps my desired capability would have been better stated, "concise prototyping". When I'm writing out a model, I don't want to have to write (or generate) a lot of boilerplate code. (or in the case of Joe's example, code whose scope is contained to a single containing component). 

> Here the concern is that you're building implicitly an instance model.

I might be misunderstanding your point, but I think that by writing the implementation without the type declaration, I'm implicitly building a feature model and _explicitly_ building an instance model. 

> I fail to see any improvements in readability when you stack also annexes in the definition.
The readability is improved in simply making the AADL more concise. An implicit type definition loses no semantic precision and reduces the amount of scrolling up and down required of the modeler. 

> Also, how would you name elements in the hierarchy, e.g. when you want to add flow ? are you applying is to the component instance or anonymous classifier? this proposal would impact name resolution.

Agreed, there would need to be some additional level of namespacing. Java does something similar with named or anonymous inner classes. 

> Note that some modelling notations also blur the distinction between component type and implementations, if you want to enter into this business of syntax reduction, we can even go further.
> But again, we have to be careful about simplification vs. strong typing.

Agreed, my hope is to simplify without losing clarity or strong typing.

**Comment by yoogx** [comment]
Yes, I meant explicitly, you're right. With some distinction with the instance model on the connections expansion.

Probably an option would be to compare with the mini language Peter prototyped a while back for a concrete syntax for the instance model. It certainly covered many of the corner cases I have in mind. 

I do not know whether it is now included in OSATE. Perhaps Joe or Peter can confirm this.

**Comment by joeseibel** [comment]
I can confirm that the textual representation of instance models is included in OSATE. Right-click on an instance file, then click "Generate Textual Instance". This will produce a separate .instance file. The textual instance contains all of the information in the standard instance model, so I'm not sure if it would be a good fit for quick prototyping. It was meant to be a read-only language.

The following is an example of a textual instance model:
```
system s_i_Instance : Nested_Subcomponents::s.i {
	process Nested_Subcomponents::ps1.i ps1 : Nested_Subcomponents::s.i:ps1 {
		out dataPort ps1_out : Nested_Subcomponents::ps1:ps1_out
		thread Nested_Subcomponents::t1 t1 : Nested_Subcomponents::ps1.i:t1 {
			out dataPort t1_out : Nested_Subcomponents::t1:t1_out
		}
	}
	process Nested_Subcomponents::ps2.i ps2 : Nested_Subcomponents::s.i:ps2 {
		in dataPort ps2_in : Nested_Subcomponents::ps2:ps2_in
		thread Nested_Subcomponents::t2 t2 : Nested_Subcomponents::ps2.i:t2 {
			in dataPort t2_in : Nested_Subcomponents::t2:t2_in
		}
	}
	complete portConnection "ps1.t1.t1_out -> ps2.t2.t2_in" : ps1.t1.t1_out -> ps2.t2.t2_in {
		ps1.t1.t1_out -> ps1.ps1_out : Nested_Subcomponents::ps1.i:ps1_conn in ps1
		ps1.ps1_out -> ps2.ps2_in : Nested_Subcomponents::s.i:across_conn in parent
		ps2.ps2_in -> ps2.t2.t2_in : Nested_Subcomponents::ps2.i:ps2_conn in ps2
	}
	som "No Modes"
}
```
You will notice that all of the components and features have references back to their declarative counterpart. In the case of subcomponents, this consists of a reference to the declarative classifier and a separate reference to the declarative subcomponent. Also, the connection instance contains references to all of the individual declarative connections which make up the connection instance.
